Login and Registration complete plus empty fragments

isabelcosta opened the pull request:
### Description

Implemented first screens and empty fragments to stat internal features development.

- Both Login and Sign Up screen show error messages when fields are empty.
- Created Bottom navigation and icon resources
- Created Empty fragments for (Home, Profile, Members, Relation, Requests)
- PopUp Menu template on Action Bar
- Splash, Login and Register do not contain Action Bar just as it is in the UI prototype

`BottomNavigationViewHelper.kt` is a hack for the bottom navigation to be still instead of having shiftingMode enabled, which is by default. I've seen this solution in many websites, [this was one of the resources.](https://stackoverflow.com/questions/40176244/how-to-disable-bottomnavigationview-shift-mode)

There's a BaseFragment for the fragments to extend from.

Fixes #2 
Fixes #3 

### Type of Change:

- Code

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?

Performed registration and login, both with valid and invalid inputs.
Received toast when a success message is received and Snackbar when an error message is received.

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation


**Code/Quality Assurance Only**
- [x] My changes generate no new warnings
- [x] Any dependent changes have been published in downstream modules

r00pal reviewed (dismissed):
Looks good to me. Just fix the codacy errors

m-murad commented on .idea/codeStyles/Project.xml:
```diff
@@ -0,0 +1,29 @@
+<component name="ProjectCodeStyleConfiguration">
```
Don't include this file

isabelcosta replied:
Can I include these 4 files on .gitignore? or even the `.idea` folder as I asked you before? to avoid these files going into git? @m-murad

m-murad commented on .idea/misc.xml:
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Don't include this file

m-murad commented on .idea/modules.xml:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Don't include this file

m-murad commented on .idea/vcs.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Don't include this file

m-murad commented on app/src/main/java/org/systers/mentorship/remote/responses/RegisterRequest.kt:
```diff
@@ -0,0 +1,12 @@
+        val password: String,
+        @SerializedName("terms_and_conditions_checked")
+        val hasAcceptedTermsAndConditions: Boolean
+)
```
blank line at the end

m-murad commented on app/src/main/java/org/systers/mentorship/remote/services/AuthService.kt:
```diff
@@ -13,4 +15,8 @@ interface AuthService {
+
+    @POST("register")
+    fun register(@Body registerRequest: RegisterRequest) : Observable<CustomResponse>
 }
```
blank line at the end

m-murad commented on app/src/main/java/org/systers/mentorship/utils/AppCompatActivityExt.kt:
```diff
@@ -0,0 +1,44 @@
+package org.systers.mentorship.utils
```
Don't create this util file. Include a single method for this in the `BaseActivity`

m-murad commented on app/src/main/java/org/systers/mentorship/view/activities/LoginActivity.kt:
```diff
@@ -41,9 +41,17 @@ class LoginActivity : BaseActivity() {
+
+        btnSignUp.setOnClickListener {
+            intent = Intent(this, SignUpActivity::class.java)
+            startActivity(intent)
```
`finish()` this activity

m-murad commented on app/src/main/java/org/systers/mentorship/view/fragments/RequestsFragment.kt:
```diff
@@ -0,0 +1,25 @@
+        private const val TAG = "RequestsFragment"
+
+    }
+}
```
add a Blank like

m-murad commented on app/src/main/java/org/systers/mentorship/view/fragments/RelationFragment.kt:
```diff
@@ -0,0 +1,24 @@
+        private const val TAG = "RelationFragment"
+
+    }
+}
```
add a Blank like

m-murad commented on app/src/main/java/org/systers/mentorship/viewmodels/SignUpViewModel.kt:
```diff
@@ -0,0 +1,73 @@
+import java.io.IOException
+import java.util.concurrent.TimeoutException
+
+/**
```
I didn't create this 😉

isabelcosta replied:
copy and paste traces ;) will remove that!

m-murad commented on app/src/main/res/layout/activity_login.xml:
```diff
@@ -1,15 +1,103 @@
+        app:srcCompat="@drawable/mentorship_system_logo" />
+
+    <android.support.design.widget.TextInputLayout
+        android:id="@+id/inputTextLoginUsername"
```
don't use long names. We are following 100 column spaces pattern in .kt files.

isabelcosta replied:
"100 column spaces pattern" I had to think a bit to understand what you meant by this. I can change that

m-murad replied:
name it something like `etUsername`

isabelcosta replied:
Ok, I'll do it like that

isabelcosta replied:
`et` for edit text and `til` for input text layout? @m-murad

m-murad commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,9 +1,41 @@
+    tools:context=".view.activities.MainActivity">
 
-</android.support.constraint.ConstraintLayout>
+    <!--<android.support.v7.widget.Toolbar-->
```
remove commented code

m-murad commented on app/src/main/res/layout/fragment_home.xml:
```diff
@@ -0,0 +1,13 @@
+        android:layout_height="match_parent"
+        android:text="@string/fragment_title_home" />
+
+</FrameLayout>
```
add blank line

m-murad commented on app/src/main/res/layout/fragment_members.xml:
```diff
@@ -0,0 +1,11 @@
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:text="@string/fragment_title_members" />
+</android.support.constraint.ConstraintLayout>
```
add blank line

m-murad commented on app/src/main/res/layout/fragment_profile.xml:
```diff
@@ -0,0 +1,12 @@
+        android:layout_height="match_parent"
+        android:text="@string/fragment_title_profile" />
+
+</android.support.constraint.ConstraintLayout>
```
add blank line

m-murad commented on app/src/main/res/layout/fragment_relation.xml:
```diff
@@ -0,0 +1,11 @@
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:text="@string/fragment_title_relation" />
+</android.support.constraint.ConstraintLayout>
```
add blank line

m-murad commented on app/src/main/res/layout/fragment_requests.xml:
```diff
@@ -0,0 +1,11 @@
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:text="@string/fragment_title_requests" />
+</android.support.constraint.ConstraintLayout>
```
add blank line

m-murad commented on app/src/main/res/menu/popup.xml:
```diff
@@ -0,0 +1,16 @@
+        android:id="@+id/menu_feedback"
+        android:title="@string/popup_menu_feedback"
+        app:showAsAction="never" />
+</menu>
```
add blank line

m-murad commented on app/src/main/java/org/systers/mentorship/view/fragments/ProfileFragment.kt:
```diff
@@ -0,0 +1,23 @@
+        fun newInstance() = ProfileFragment()
+        private const val TAG = "ProfileFragment"
+    }
+}
```
add blank line

m-murad commented on app/src/main/java/org/systers/mentorship/view/fragments/MembersFragment.kt:
```diff
@@ -0,0 +1,24 @@
+        private const val TAG = "MembersFragment"
+
+    }
+}
```
add blank line

m-murad commented on app/src/main/java/org/systers/mentorship/view/fragments/HomeFragment.kt:
```diff
@@ -0,0 +1,25 @@
+        private const val TAG = "HomeFragment"
+
+    }
+}
```
add blank line

m-murad commented on app/src/main/java/org/systers/mentorship/view/fragments/BaseFragment.kt:
```diff
@@ -0,0 +1,42 @@
+     * @return string of the title of the fragment
+     */
+    protected abstract fun getTitleResourceId() : Int
+}
```
add blank line

m-murad commented on app/src/main/java/org/systers/mentorship/view/activities/SignUpActivity.kt:
```diff
@@ -0,0 +1,132 @@
+                if (successful) {
+                    Toast.makeText(this, signUpViewModel.message, Toast.LENGTH_LONG)
+                            .show()
+                    onBackPressed()
```
Don't use `onBackPressed()`. Finish the previous activity and start the next activity using an intent.

Also, include a button in this layout to go to LoginActivity.

isabelcosta replied:
its included, I'll show the UI after the review fixes are complete

m-murad commented on app/src/main/java/org/systers/mentorship/view/activities/SignUpActivity.kt:
```diff
@@ -0,0 +1,132 @@
+     * @param areTermsAndConditionsChecked true if the terms and conditions is checked, false if otherwise
+     * @return true if all field values are valid, false if otherwise
+     */
+    private fun areValidInputs(name: String, username: String, email: String,
```
Change name to `areInputsValid`or `validateInputs`

m-murad commented on app/src/main/java/org/systers/mentorship/view/activities/SignUpActivity.kt:
```diff
@@ -0,0 +1,132 @@
+    private fun navigateToLogin() {
+        onBackPressed()
+    }
+}
```
Blank line

m-murad commented on app/src/main/java/org/systers/mentorship/view/fragments/HomeFragment.kt:
```diff
@@ -0,0 +1,25 @@
+
+    companion object {
+        fun newInstance() = HomeFragment()
+        private const val TAG = "HomeFragment"
```
TAG = `HomeFragment::class.java.simpleName`

Change in other fragments aswell

m-murad commented on app/src/main/java/org/systers/mentorship/view/fragments/HomeFragment.kt:
```diff
@@ -0,0 +1,25 @@
+    companion object {
+        fun newInstance() = HomeFragment()
+        private const val TAG = "HomeFragment"
+
```
Avoid blank lines

m-murad commented on app/src/main/java/org/systers/mentorship/view/fragments/MembersFragment.kt:
```diff
@@ -0,0 +1,24 @@
+    companion object {
+        fun newInstance() = MembersFragment()
+        private const val TAG = "MembersFragment"
+
```
avoid blank lines

m-murad commented:
When your PR changes UI make sure that you include screenshots here.

m-murad commented on app/src/main/res/values/strings.xml:
```diff
@@ -2,9 +2,38 @@
+    <string name="username">Username</string>
+    <string name="password">Password</string>
+    <string name="confirm_password">Confirm password</string>
+    <string name="terms_and_conditions_acceptance_statement"><![CDATA[I agree and accept the Terms & Conditions of the Mentorship System]]></string>
```
I agree to accept...

m-murad commented on app/src/main/res/values/strings.xml:
```diff
@@ -2,9 +2,38 @@
     <string name="logging_successful">Logging successful</string>
+    <string name="registration_successful">Registration successful. Please confirm your email!</string>
+    <string name="hint_login_username_or_email">Username or Email</string>
+    <string name="hint_login_password">Password</string>
```
don't use `hint_`
because the string can be displayed anywhere (as hint or even as text).

Also, remove duplicates in this case `hint_login_password` and `password`. Just keep `password `

m-murad commented on app/src/main/res/values/strings.xml:
```diff
@@ -2,9 +2,38 @@
+    <string name="password">Password</string>
+    <string name="confirm_password">Confirm password</string>
+    <string name="terms_and_conditions_acceptance_statement"><![CDATA[I agree and accept the Terms & Conditions of the Mentorship System]]></string>
+    <string name="error_not_matching_passwords">Passwords are not matching!</string>
```
Passwords didn't match!

m-murad commented on app/src/main/java/org/systers/mentorship/view/activities/LoginActivity.kt:
```diff
@@ -52,4 +60,28 @@ class LoginActivity : BaseActivity() {
+     */
+    private fun areValidInputs(username: String, password: String) : Boolean {
+        var isValidInput = true
+        if (username.isEmpty()) {
```
use `isNullOrBlank()`

m-murad commented on app/src/main/java/org/systers/mentorship/view/activities/LoginActivity.kt:
```diff
@@ -52,4 +60,28 @@ class LoginActivity : BaseActivity() {
+        } else {
+            inputTextLoginUsername.error = null
+        }
+        if (password.isEmpty()) {
```
use `isNullOrBlank`

m-murad commented on app/src/main/java/org/systers/mentorship/view/activities/LoginActivity.kt:
```diff
@@ -52,4 +60,28 @@ class LoginActivity : BaseActivity() {
+    private fun areValidInputs(username: String, password: String) : Boolean {
+        var isValidInput = true
+        if (username.isEmpty()) {
+            inputTextLoginUsername.error = getString(R.string.error_empty_username)
```
before setting error always request for focus using `requestFocus()`

isabelcosta commented:
About the UI screenshots, I agree! I'll do that on the PR description

isabelcosta commented:
Login Screen (without input errors): 

![image](https://user-images.githubusercontent.com/11148726/43465486-5afa6884-94d5-11e8-9ada-7c836a718cd9.png)

Login Screen (with input errors):

![image](https://user-images.githubusercontent.com/11148726/43465535-73f6ac76-94d5-11e8-91f8-4bd8d3ac6ca9.png)

isabelcosta commented:
Sign Up (terms unchecked):

![image](https://user-images.githubusercontent.com/11148726/43465589-91f69b5a-94d5-11e8-9583-216236c9c946.png)

Sign Up (terms checked):

![image](https://user-images.githubusercontent.com/11148726/43465640-a77165d2-94d5-11e8-931c-d4ceded33da3.png)

Sign Up (empty fields)... has scroll:

![image](https://user-images.githubusercontent.com/11148726/43465683-bf367f0e-94d5-11e8-912c-e1f1792683bb.png)

isabelcosta commented:
Main Activtity:

![image](https://user-images.githubusercontent.com/11148726/43465781-fc695608-94d5-11e8-8f1e-5803285792ed.png)

PopUp Menu:

![image](https://user-images.githubusercontent.com/11148726/43465794-050fff3c-94d6-11e8-9e5d-376d65a7b9c9.png)

isabelcosta commented:
I'm working on the Codacy warnings

isabelcosta commented:
Issue https://app.codacy.com/app/systers-opensource/mentorship-android/file/20368409301/issues/source?bid=7425875&fileBranchId=8213215#l32 if I do what Codacy says Android studio will tell me I have an error and make me put as it is now

isabelcosta commented:
@systers/maintainers-mentorship-android can you give it another review? I fixed @m-murad comments. I addressed the Codacy comments I could fix @roopalJazz and @Dilu9218

m-murad commented on app/src/main/java/org/systers/mentorship/view/activities/SignUpActivity.kt:
```diff
@@ -0,0 +1,119 @@
+            if (validateDetails()) {
+                val requestData = RegisterRequest(name, username, password, email, true)
+                signUpViewModel.register(requestData)
+                showProgressDialog(getString(R.string.logging_in))
```
change this message

isabelcosta replied:
Ok will fix this!

isabelcosta replied:
can be Signing up...?

m-murad replied:
yes

m-murad commented on app/src/main/java/org/systers/mentorship/view/activities/BaseActivity.kt:
```diff
@@ -51,4 +52,14 @@ open class BaseActivity: AppCompatActivity() {
     }
+
+    /**
+     * The `fragment` is added to the container view with id `frameId`. The operation is
```
my bad, please update the docs

isabelcosta committed:
feat: login, register and main activity

m-murad commented:
LGTM
